src/storage/txn/commands/acquire_pessimistic_lock_resumed.rs

```diff
@@ -0,0 +1,211 @@
+    fn process_write(self, snapshot: S, context: WriteContext<'_, L>) -> Result<WriteResult> {
+        let mut modifies = vec![];
+        let mut txn = None;
+        let mut reader: Option<SnapshotReader<S>> = None;
+
+        let mut written_rows = 0;
+        let mut res = PessimisticLockResults::with_capacity(self.items.len());
+        let mut encountered_locks = vec![];
+        let need_old_value = context.extra_op == ExtraOp::ReadOldValue;
+        let mut old_values = OldValues::default();
+
+        let mut new_locked_keys = Vec::with_capacity(self.items.len());
+
+        for item in self.items.into_iter() {
+            let ResumedPessimisticLockItem {
+                key,
+                should_not_exist,
+                params,
+                lock_wait_token,
+            } = item;
+
+            // TODO: Refine the code for rebuilding txn state.
+            if txn
+                .as_ref()
+                .map_or(true, |t: &MvccTxn| t.start_ts != params.start_ts)
+            {
+                if let Some(txn) = txn {
+                    if !txn.is_empty() {
+                        modifies.extend(txn.into_modifies());
+                    }
+                }
+                txn = Some(MvccTxn::new(
+                    params.start_ts,
+                    context.concurrency_manager.clone(),
+                ));
+                // TODO: Is it possible to reuse the same reader but change the start_ts stored
+                // in it?
+                if let Some(mut reader) = reader {
+                    context.statistics.add(&reader.take_statistics());
+                }
+                reader = Some(SnapshotReader::new_with_ctx(
+                    params.start_ts,
+                    snapshot.clone(),
+                    &self.ctx,
+                ));
+            }
+            let txn = txn.as_mut().unwrap();
+            let reader = reader.as_mut().unwrap();
+
+            match acquire_pessimistic_lock(
+                txn,
+                reader,
+                key.clone(),
+                &params.primary,
+                should_not_exist,
+                params.lock_ttl,
+                params.for_update_ts,
+                params.return_values,
+                params.check_existence,
+                params.min_commit_ts,
+                need_old_value,
+                params.lock_only_if_exists,
+                true,
+            ) {
+                Ok((key_res, old_value)) => {
+                    res.push(key_res);
+                    new_locked_keys.push((params.start_ts, key.clone()));
+                    if old_value.resolved() {
+                        let key = key.append_ts(txn.start_ts);
+                        // MutationType is unknown in AcquirePessimisticLock stage.
+                        let mutation_type = None;
+                        old_values.insert(key, (old_value, mutation_type));
+                    }
```

Comment:
Seems we can extract this common code from acquire_pessimistic_lock, acquire_pessimistic_lock_resumed and prewrite.